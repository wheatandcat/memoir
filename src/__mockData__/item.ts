import { ItemQuery } from 'queries/api/index';
type Item = NonNullable<ItemQuery['item']>;

export const item = (option?: Partial<Item>): Item => ({
  id: '1',
  title: '買い物',
  categoryID: 9,
  date: '2021-01-01T00:00:00+09:00',
  like: false,
  dislike: false,
  createdAt: '2021-01-01T00:00:00+09:00',
  updatedAt: '2021-01-01T00:00:00+09:00',
  ...option,
});

export const items = (): Item[] => [
  item(),
  item({
    id: '2',
    title: '本を読む',
    categoryID: 2,
    date: '',
    dislike: false,
    like: false,
  }),
];
