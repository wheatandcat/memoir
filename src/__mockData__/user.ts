import { User } from 'queries/api/index';

export const user = (option?: Partial<User>): User => ({
  id: '1',
  displayName: 'test',
  image: '',
  updatedAt: '2021-01-01T00:00:00+09:00',
  createdAt: '2021-01-01T00:00:00+09:00',
  ...option,
});

export const users = (): User[] => [
  {
    id: 'test1',
    displayName: 'suzuki1',
    image: 'https://placehold.jp/150x150.png',
    updatedAt: '2021-01-01T00:00:00+09:00',
    createdAt: '2021-01-01T00:00:00+09:00',
  },
  {
    id: 'test2',
    displayName: 'suzuki2',
    image: 'https://placehold.jp/150x150.png',
    updatedAt: '2021-01-01T00:00:00+09:00',
    createdAt: '2021-01-01T00:00:00+09:00',
  },
  {
    id: 'test3',
    displayName: 'suzuki4',
    image: 'https://placehold.jp/150x150.png',
    updatedAt: '2021-01-01T00:00:00+09:00',
    createdAt: '2021-01-01T00:00:00+09:00',
  },
  {
    id: 'test4',
    displayName: 'suzuki4',
    image: 'https://placehold.jp/150x150.png',
    updatedAt: '2021-01-01T00:00:00+09:00',
    createdAt: '2021-01-01T00:00:00+09:00',
  },
  {
    id: 'test5',
    displayName: 'suzuki5',
    image: 'https://placehold.jp/150x150.png',
    updatedAt: '2021-01-01T00:00:00+09:00',
    createdAt: '2021-01-01T00:00:00+09:00',
  },
];
