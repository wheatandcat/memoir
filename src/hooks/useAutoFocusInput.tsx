import { useRef } from 'react';
import { TextInput } from 'react-native';
import { useFocusEffect } from '@react-navigation/native';

export default function useAutoFocusInput(autoFocus: boolean) {
  const ref = useRef<TextInput>(null);
  const timeout = useRef<number | null>(null);

  useFocusEffect(() => {
    if (autoFocus) {
      timeout.current = Number(setTimeout(() => ref.current?.focus(), 300));
    }

    return () => {
      if (timeout.current) {
        clearTimeout(timeout.current);
      }
    };
  });

  return {
    ref,
    autoFocus: false,
  };
}
