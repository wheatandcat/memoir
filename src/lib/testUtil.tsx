import React from 'react';
import { GraphQLHandler, GraphQLRequest } from 'msw';
import {
  ApolloClient,
  ApolloProvider,
  createHttpLink,
  InMemoryCache,
} from '@apollo/client';
import { render } from '@testing-library/react-native';
import { server } from 'mocks/server';
import fetch from 'cross-fetch';

const link = createHttpLink({
  uri: 'http://localhost:8080/query',
  fetch,
  credentials: 'same-origin',
});

const client = new ApolloClient({
  link,
  uri: 'http://localhost:8080/query',
  cache: new InMemoryCache(),
});

export const testRenderer =
  (children: React.ReactNode) =>
  (responseOverride?: GraphQLHandler<GraphQLRequest<never>>) => {
    if (responseOverride) {
      server.use(responseOverride);
    }
    return render(<ApolloProvider client={client}>{children}</ApolloProvider>);
  };
