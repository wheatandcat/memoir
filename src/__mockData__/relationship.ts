import type { RelationshipsQuery } from "queries/api/index";
import { user } from "./user";

export type Relationship = NonNullable<
  EdgesNode<RelationshipsQuery["relationships"]>
>;

export const relationship = (option?: Partial<Relationship>): Relationship => ({
  id: "1",
  followedId: "test",
  followerId: "test",
  user: user(),
  createdAt: "2021-01-01T00:00:00+09:00",
  updatedAt: "2021-01-01T00:00:00+09:00",
  ...option,
});

export const relationships = (): Relationship[] => [
  relationship(),
  relationship({
    id: "2",
    user: user(),
  }),
  relationship({
    id: "3",
    user: user(),
  }),
  relationship({
    id: "4",
    user: user(),
  }),
];
