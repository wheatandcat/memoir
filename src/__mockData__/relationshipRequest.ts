import { RelationshipRequestsQuery } from 'queries/api/index';
import { user } from './user';

export type RelationshipRequest = NonNullable<
  EdgesNode<RelationshipRequestsQuery['relationshipRequests']>
>;

export const relationshipRequest = (
  option?: Partial<RelationshipRequest>
): RelationshipRequest => ({
  id: '1',
  followedId: 'test',
  followerId: 'test',
  user: user(),
  createdAt: '2021-01-01T00:00:00+09:00',
  updatedAt: '2021-01-01T00:00:00+09:00',
  ...option,
});

export const relationshipRequests = (): RelationshipRequest[] => [
  relationshipRequest(),
  relationshipRequest({
    id: '2',
    user: user(),
  }),
  relationshipRequest({
    id: '3',
    user: user(),
  }),
  relationshipRequest({
    id: '4',
    user: user(),
  }),
];
