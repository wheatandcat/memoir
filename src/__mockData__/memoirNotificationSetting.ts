import { State as MemoirNotificationSettingState } from 'hooks/useMemoirNotificationSetting';

export const memoirNotificationSetting = (
  option?: Partial<MemoirNotificationSettingState>
): MemoirNotificationSettingState => ({
  dayOfWeek: 0,
  hours: 0,
  minutes: 0,
  notification: false,
  ...option,
});
