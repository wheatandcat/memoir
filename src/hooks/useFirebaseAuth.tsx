import * as AppleAuthentication from 'expo-apple-authentication';
import * as WebBrowser from 'expo-web-browser';
import * as Google from 'expo-auth-session/providers/google';
import { v4 as uuidv4 } from 'uuid';
import { ResponseType } from 'expo-auth-session';
import * as Crypto from 'expo-crypto';
import { useRecoilValueLoadable, useRecoilState } from 'recoil';
import { useCallback, useEffect, useState } from 'react';
import { Alert } from 'react-native';
import firebase from 'lib/system/firebase';
import 'lib/firebase';
import { storageKey, getItem, removeItem } from 'lib/storage';
import { existAuthUserID } from 'store/selectors';
import { authUserState, userState } from 'store/atoms';
import Auth from 'lib/auth';
import {
  useCreateAuthUserMutation,
  CreateAuthUserMutationVariables,
  useExistAuthUserLazyQuery,
  useUserLazyQuery,
} from 'queries/api/index';

const auth = new Auth();

WebBrowser.maybeCompleteAuthSession();

const nonceGen = (length: number) => {
  let result = '';
  let characters =
    'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
  let charactersLength = characters.length;
  for (let i = 0; i < length; i++) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength));
  }
  return result;
};

export type UseFirebaseAuth = ReturnType<typeof useFirebaseAuth>;

const useFirebaseAuth = (errorCallback?: () => void) => {
  const [setup, setSetup] = useState(false);
  const authUserID = useRecoilValueLoadable(existAuthUserID);
  const [authUser, setAuthUser] = useRecoilState(authUserState);
  const [user, setUser] = useRecoilState(userState);

  useEffect(() => {
    if (user.id) {
      // Auth認証後はuserの設定が完了した際にsetupを完了にする
      setSetup(true);
    }
  }, [user.id]);

  const [getUser] = useUserLazyQuery({
    onCompleted: (data) => {
      setUser((s) => ({
        ...s,
        id: data?.user?.id || '',
        userID: data?.user?.id || '',
        displayName: data?.user?.displayName || '',
        image: data?.user?.image || '',
      }));
    },
  });

  const [createAuthUserMutation] = useCreateAuthUserMutation({
    async onCompleted(data) {
      const id = data.createAuthUser.id;

      setUser({
        id,
        userID: '',
        displayName: '',
        image: '',
      });

      setSetup(true);
    },
    async onError() {
      // エラーになった場合はログアウトさせる
      Alert.alert('エラー', 'ログインに失敗した');
      onLogout();
    },
  });

  const [getExistAuthUser] = useExistAuthUserLazyQuery({
    onCompleted: (data) => {
      if (data.existAuthUser.exist === false) {
        const u = uuidv4();
        const variables: CreateAuthUserMutationVariables = {
          input: {
            id: u,
            isNewUser: true,
          },
        };

        createAuthUserMutation({
          variables,
        });
      } else {
        // ユーザー情報を設定
        getUser();
      }
    },
  });

  const [request, response, promptAsync] = Google.useIdTokenAuthRequest({
    responseType: ResponseType.IdToken,
    clientId: process.env.GOOGLE_CLIENT_ID,
    expoClientId: process.env.EXPO_GOOGLE_CLIENT_ID,
  });

  const onGoogleLogin = useCallback(() => {
    promptAsync();
  }, [promptAsync]);

  const setSession = useCallback(
    async (refresh = false) => {
      const idToken = await auth.setSession(refresh);

      getExistAuthUser();

      if (idToken) {
        const authUID = await getItem(storageKey.AUTH_UID_KEY);
        setAuthUser({
          uid: authUID,
        });
      }

      return idToken;
    },
    [setAuthUser, getExistAuthUser]
  );

  const firebaseLogin = useCallback(
    async (credential: firebase.auth.OAuthCredential) => {
      const data = await firebase
        .auth()
        .signInWithCredential(credential)
        .catch((error: any) => {
          console.log(error);
        });

      console.log(data);

      const ok = await setSession(true);

      return ok;
    },
    [setSession]
  );

  useEffect(() => {
    if (response?.type === 'success') {
      const { id_token } = response.params;
      const credential = firebase.auth.GoogleAuthProvider.credential(id_token);
      firebaseLogin(credential);
    } else if (response?.type === 'error') {
      console.log('error:', response);
      Alert.alert('ログインに失敗しました');
      errorCallback?.();
    }
  }, [response, firebaseLogin, errorCallback]);

  const onAppleLogin = useCallback(async () => {
    const nonce = nonceGen(32);
    const digestedNonce = await Crypto.digestStringAsync(
      Crypto.CryptoDigestAlgorithm.SHA256,
      nonce
    );

    try {
      const result = await AppleAuthentication.signInAsync({
        requestedScopes: [
          AppleAuthentication.AppleAuthenticationScope.FULL_NAME,
          AppleAuthentication.AppleAuthenticationScope.EMAIL,
        ],
        nonce: digestedNonce,
      });
      const provider = new firebase.auth.OAuthProvider('apple.com');
      const credential = provider.credential({
        idToken: result.identityToken || '',
        rawNonce: nonce,
      });

      firebaseLogin(credential);
    } catch (e) {
      console.log('error:', e);
      Alert.alert('ログインに失敗しました');
      errorCallback?.();
    }
  }, [firebaseLogin, errorCallback]);

  const onLogout = useCallback(async () => {
    await auth.logout();
    await removeItem(storageKey.USER_ID_KEY);

    setAuthUser({
      uid: null,
    });
    setUser({ id: null, userID: '', displayName: '', image: '' });
  }, [setAuthUser, setUser]);

  useEffect(() => {
    if (authUser.uid) {
      return;
    }
    if (authUserID.state === 'hasValue') {
      if (authUserID.contents) {
        setAuthUser({ uid: authUserID.contents });
      }
    }
  }, [authUserID, setAuthUser, authUser.uid]);

  useEffect(() => {
    const unsubscribe = firebase.auth().onAuthStateChanged((user) => {
      if (user) {
        setSession(true);
      } else {
        // ログアウトした時
        setAuthUser({
          uid: null,
        });
        setUser({ id: null, userID: '', displayName: '', image: '' });
        setSetup(true);
      }
    });

    return () => unsubscribe();
  }, [setUser, setAuthUser, setSession]);

  return {
    setupAuth: setup,
    request,
    onAppleLogin,
    onGoogleLogin,
    onLogout,
  };
};

export default useFirebaseAuth;
