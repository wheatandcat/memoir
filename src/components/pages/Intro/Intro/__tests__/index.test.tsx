import React from 'react';
import { testRenderer } from 'lib/testUtil';
import { screen } from '@testing-library/react-native';
import * as useMemoirNotificationSetting from 'hooks/useMemoirNotificationSetting';
import { memoirNotificationSetting } from '__mockData__/memoirNotificationSetting';
import IndexPage from '../';

const propsData = () => ({ onFinish: jest.fn() });

describe('components/pages/Intro/Intro/index.tsx', () => {
  beforeEach(() => {
    jest
      .spyOn(useMemoirNotificationSetting, 'default')
      .mockImplementation((): any => ({
        ...memoirNotificationSetting(),
        loading: false,
        onSave: jest.fn(),
      }));
  });

  it('正常にrenderすること', async () => {
    testRenderer(<IndexPage {...propsData()} />)();
    expect(screen.findAllByText('SKIP')).toBeTruthy();
  });
});
