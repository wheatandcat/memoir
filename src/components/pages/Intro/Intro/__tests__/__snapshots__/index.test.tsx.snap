// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pages/Intro/Intro/index.tsx 正常にrenderすること 1`] = `
<Memo(Connected)
  onFinish={[MockFunction]}
/>
`;
